website/server/libs/taskManager.js

```diff
@@ -0,0 +1,170 @@
+    validateBeforeSave: false,
+  }));
+
+  toSave.unshift((challenge || user).save());
+
+  let tasks = await Bluebird.all(toSave);
+  tasks.splice(0, 1); // Remove user, challenge, or group promise
+  return tasks;
+}
+
+/**
+ * Gets tasks for a user, challenge or group.
+ *
+ * @param  req  The Express req variable
+ * @param  res  The Express res variable
+ * @param  options  Should be a user, user and challenge to get challenge tasks or user and group to get group tasks
+ * @return The tasks found
+ */
+export async function getTasks (req, res, options = {}) {
+  let {
+    user,
+    challenge,
+    group,
+  } = options;
+
+  let query = {userId: user._id};
+
+  if (challenge) {
+    query =  {'challenge.id': challenge.id, userId: {$exists: false}};
+  } else if (group) {
+    query =  {'group.id': group._id};
+  }
+
+  let type = req.query.type;
+
+  if (type) {
+    if (type === 'todos') {
+      query.completed = false; // Exclude completed todos
+      query.type = 'todo';
+    } else if (type === 'completedTodos' || type === '_allCompletedTodos') { // _allCompletedTodos is currently in BETA and is likely to be removed in future
+      let limit = 30;
+
+      if (type === '_allCompletedTodos') {
+        limit = 0; // no limit
+      }
+      query = Tasks.Task.find({
+        userId: user._id,
+        type: 'todo',
+        completed: true,
+      }).limit(limit).sort({
+        dateCompleted: -1,
+      });
+    } else {
+      query.type = type.slice(0, -1); // removing the final "s"
+    }
+  } else {
+    query.$or = [ // Exclude completed todos
+      {type: 'todo', completed: false},
+      {type: {$in: ['habit', 'daily', 'reward']}},
+    ];
+  }
+
+  let tasks = await Tasks.Task.find(query).exec();
+
+  // Order tasks based on tasksOrder
+  if (type && type !== 'completedTodos' && type !== '_allCompletedTodos') {
+    let order = (challenge || user).tasksOrder[type];
+    let orderedTasks = new Array(tasks.length);
+    let unorderedTasks = []; // what we want to add later
+
+    tasks.forEach((task, index) => {
+      let taskId = task._id;
+      let i = order[index] === taskId ? index : order.indexOf(taskId);
+      if (i === -1) {
+        unorderedTasks.push(task);
+      } else {
+        orderedTasks[i] = task;
+      }
+    });
+
+    // Remove empty values from the array and add any unordered task
+    orderedTasks = _.compact(orderedTasks).concat(unorderedTasks);
+    return orderedTasks;
+  } else {
+    return tasks;
+  }
+}
+
+// Takes a Task document and return a plain object of attributes that can be synced to the user
+
+export function syncableAttrs (task) {
+  let t = task.toObject(); // lodash doesn't seem to like _.omit on Document
+  // only sync/compare important attrs
+  let omitAttrs = ['_id', 'userId', 'challenge', 'history', 'tags', 'completed', 'streak', 'notes', 'updatedAt', 'group', 'assignedUsers', 'linkedTaskId'];
```

Comment:
Okay, but I think we should think more generally though. What if we want assigned users from challenges or something else in the future. The linkedTaskID is an example. We didn't generalize the challenge linked id, now we have two fields that are essentially the same.